Python/Product/Analysis/Intellisense/LinePreservingCodeReplacer.cs

```diff
@@ -0,0 +1,159 @@
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.PythonTools.Parsing;
+
+namespace Microsoft.PythonTools.Intellisense {
+    /// <summary>
+    /// Replaces a range of text with new text attempting only modifying lines which changed 
+    /// and doing so in a single edit.
+    /// </summary>
+    class LinePreservingCodeReplacer {
+        private readonly string _newLine, _oldCode, _newCode;
+        private readonly IReadOnlyList<LineInfo> _newLines, _oldLines;
+
+        private LinePreservingCodeReplacer(string oldCode, string newCode, string newLine) {
+            _newLine = newLine;
+            _oldCode = oldCode;
+            _newCode = newCode;
+            _oldLines = LineInfo.SplitLines(oldCode).ToArray();
+            _newLines = LineInfo.SplitLines(newCode).ToArray();
+        }
+
+        private string GetOldText(int line) {
+            return _oldCode.Substring(_oldLines[line].Start, _oldLines[line].Length);
+        }
+
+        private string GetNewText(int line) {
+            return _newCode.Substring(_newLines[line].Start, _newLines[line].Length);
+        }
+
+        public static IReadOnlyList<DocumentChange> Replace(string oldCode, string newCode, string newLine = "\r\n") {
+            return new LinePreservingCodeReplacer(oldCode, newCode, newLine).ReplaceCode();
+        }
+
+        public IReadOnlyList<DocumentChange> ReplaceCode() {
+            var edits = new List<DocumentChange>();
+            var oldLineMapping = new Dictionary<string, List<int>>();   // line to line #
+            for (int i = 0; i < _oldLines.Count; i++) {
+                List<int> lineInfo;
+                if (!oldLineMapping.TryGetValue(GetOldText(i), out lineInfo)) {
+                    oldLineMapping[GetOldText(i)] = lineInfo = new List<int>();
+                }
+                lineInfo.Add(i);
+            }
+
+            int curOldLine = 0;
+            for (int curNewLine = 0; curOldLine < _oldLines.Count && curNewLine < _newLines.Count; curOldLine++) {
+                if (GetOldText(curOldLine) == GetNewText(curNewLine)) {
+                    curNewLine++;
+                    continue;
+                }
+
+                bool replaced = false;
+                // replace starts on this line, figure out where it ends...
+                int startNewLine = curNewLine;
+                for (curNewLine += 1; curNewLine < _newLines.Count; curNewLine++) {
```

Comment:
Sorry, my original comment was incorrect.

But there are still some cases when a simple change can result in a replacement of a big chunk.
For instance, there is a subsequence of `n+1` lines `A..NA`, and user adds a new line to the beginning of it: `ZA..NA`. Loop will start with `curNewLine == 1`, `GetNewText(curNewLine)` will return line `A`, `lines` will contain `0` and `n`, and as a result `A..NA` will be replaced with `ZA`. The remaining comparison will most likely end up with full replacement.

I'm not sure how often such thing may happen and how critical is it at all.

Reply:
File a new issue for this - I'm not going to start messing with this algorithm right now.

Reply:
Ok. https://github.com/Microsoft/PTVS/issues/3566